In [14]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import r2_score

## Part 1. Loading the dataset

In [15]:
# Using pandas load the dataset (load remotely, not locally)
data = pd.read_csv("science_data_large.csv")

# Output the first 15 rows of the data
print(data.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print(data.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [16]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data[['Temperature °C', 'Mols KCL']]
y = data['Size nm^3']
#Temperature °C Mols KCL    Size nm^3
#959	        786         1536707.314
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, random_state=42)

## Part 3. Perform a Linear Regression

In [17]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample_datapoint = [[959, 786]]  # Expected output: 1536707.314
predicted_size = model.predict(sample_datapoint)
print(predicted_size)

score = model.score(X_test, y_test)
print("Score:", score)
# Report on the score for that model, in your own words (markdown, not code) explain what the score means

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

print("Coefficients:", coefficients)
print("Intercept:", intercept)

[1232941.2530981]
Score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


The score of the model indicates how well the temperature and mols of KCl explain the variance in the size of the slime. A score close to 1 suggests a good fit, while a score close to 0 suggests a poor fit.


Equation: $h(x) = \text{intercept} + \text{coefficient}_1 \cdot \text{Temperature} + \text{coefficient}_2 \cdot \text{Mols}$

$h(x) = -409391.47958340764 + 866.14641337 \cdot \text{Temperature} + 1032.69506649 \cdot \text{Mols}$

## Part 4. Use Cross Validation

In [18]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(model, X, y, cv = 5)
print("Cross-Validation Scores:", scores)
print("Mean Cross-Validation Score:", scores.mean())
# Report on their finding and their significance

Cross-Validation Scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean Cross-Validation Score: 0.8568167899144437


Report on Findings and Their Significance

1. Cross-Validation Scores:
   - The individual scores from the cross-validation are as follows: 
     - Fold 1: 0.8392
     - Fold 2: 0.8705
     - Fold 3: 0.8587
     - Fold 4: 0.8720
     - Fold 5: 0.8436
   - These scores represent the model's score for each of the five different splits of the dataset. 

2. Mean Cross-Validation Score:
   - The mean cross-validation score is approximately 0.8568. This indicates that, on average, the model explains about 85.68% of the variance in the size of the slime based on the features. 

3. Significance of Findings:
   - The high mean score (>0.80) suggests that the model is performing well in predicting the size of the slime. 
   - The scores are relatively consistent across the folds indicating that the model is somewhat stable and not overly reliant on any single subset of the data.

4. Overall Results:
   - The values do not vary drastically, with the lowest score being 0.8392 and the highest being 0.8720, which reflects that the model is generally robust. This consistency across different data splits reinforces the reliability of the model's predictions.

## Part 5. Using Polynomial Regression

In [19]:
# Augment the dataset using PolynomialFeatures
poly = PolynomialFeatures(degree=2)
X_train_poly = poly.fit_transform(X_train)
X_test_poly = poly.transform(X_test)

# Train the model using the augmented dataset
model_poly = LinearRegression()
model_poly.fit(X_train_poly, y_train)

# Predict on the test set
y_pred_poly = model_poly.predict(X_test_poly)

# Calculate and print the R^2 score for the polynomial regression
poly_r2_score = r2_score(y_test, y_pred_poly)
print("Score:", poly_r2_score)

# Extract the coefficients and intercept
coefficients_poly = model_poly.coef_
intercept_poly = model_poly.intercept_

print("Coefficients:", coefficients_poly)
print("Intercept:", intercept_poly)

# Report on the metrics and output the resultant equation as you did in Part 3.

Score: 1.0
Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0477105863392353e-05


Report on Findings and Their Significance

$R^2$ Score:
   - The model has achieved a perfect score (1.0). The model fits the dataset perfectly (probably due to its small size).

Equation: $h(x) = \text{intercept} + \text{coefficient}_0 + \text{coefficient}_1 \cdot \text{Temperature} + \text{coefficient}_2 \cdot \text{Mols} + \text{coefficient}_3 \cdot \text{Temperature}^2 + \text{coefficient}_4 \cdot \text{Temperature} \cdot \text{Mols} + \text{coefficient}_5 \cdot \text{Mols}^2$

$h(x) = \text{2.0477105863392353e-05} + 0 + 12 \cdot \text{Temperature} + (-1.27195488e-07) \cdot \text{Mols} + 1.26494371e-11 \cdot \text{Temperature}^2 + 2 \cdot \text{Temperature} \cdot \text{Mols} +  2.85714287e-02 \cdot \text{Mols}^2$

